File: app/calculator/Calculator.scala
package calculator

import datamodel.Lesson

import scala.collection.mutable
import scala.util.Sorting

/**
 * Provide methods to calculate (possible) schedule layouts
 *
 * @author justusadam
 *
 */
object Calculator {

  /**
   * Calculate the possible layout for the provided list of lessons
   *
   * @param lessons list of lessons to allocate
   * @return List of Maps of lessons mapped to times
   */
  def calc(lessons: List[Lesson]) : List[Map[(Int, Int), Lesson]] = {

    val mapped_lessons: Map[String, List[Lesson]] = lessons groupBy (_.subject)

    val sort_func:(Lesson) => Int = (a: Lesson) => a.weight

    val sorted_lessons = mapped_lessons mapValues (
      x =>
        Sorting.stableSort(
          x,
          sort_func
        ).toList
      )

    val (min_list_primer:List[Lesson], lists_values: List[(String, List[Lesson])]) = (for ((t, x :: xs) <- sorted_lessons) yield (x, (t, xs))).unzip
    val lists = lists_values.toMap

    val x :: min_list = Sorting.stableSort(min_list_primer, sort_func).toList

    calc_step(x, lists, Map[(Int, Int), Lesson](), min_list)

  }

  def calc_step(
                 x: Lesson,
                 lists: Map[String, List[Lesson]],
                 map: Map[(Int, Int), Lesson],
                 min_list: List[Lesson]

                 ): List[Map[(Int, Int), Lesson]] = {

    def reduce_lists(subject: String, map: Map[(Int, Int), Lesson]) = {
      lists.get(subject) map {
        case Nil => Nil
        case c :: cs =>
          calc_step(c, lists.updated(subject, cs), map, min_list)
      }
    }

    map.get(x.time) match {
      case None =>

        val new_map = map + ((x.time, x))

        min_list match {
          case Nil => List(new_map)
          case c :: cs =>
            calc_step(c, lists, new_map, cs)
        }

      case Some(old) =>
        val r1 = reduce_lists(x.subject, map) getOrElse Nil
        val r2 = reduce_lists(old.subject, map + ((x.time, x))) getOrElse Nil

        r1 ++ r2
    }
  }
}

File: src/Calculator.hs
{-|
Module      : $Header$
Description : Calculate schedules
Copyright   : (c) Justus Adam, 2015
License     : LGPL-3
Maintainer  : development@justusadam.com
Stability   : experimental
Portability : POSIX

This module provides functions for calculating possibilities for an ideal
schedule layout from weighted Lessons as well as providing functions for
converting them into readable/printable format.
-}
module Calculator (
  calc,
  formatSchedule,
  totalWeight,
  time,
  Lesson (..),
  Timeslot (..)
  ) where

import           Data.List     as List (intercalate, sortBy)
import qualified Data.Map.Lazy as Map (Map, empty, foldl, fromList,
                                       fromListWith, insert, lookup, map,
                                       toList)
import qualified Data.Ord      as Ord (comparing)
import           Text.Printf   (printf)


daysPerWeek = 7
slotsPerDay = 7


data Lesson = Lesson {
  timeslot :: Int,
  day      :: Int,
  weight   :: Int,
  subject  :: String
} deriving (Show)


-- |type Alias for readability
-- maps lessons to their respective subject
type MappedLessons  = Map.Map String [Lesson]
-- |type Alias for readability
-- (Slot, Day)
type Timeslot       = (Int, Int)
-- |type Alias for readability
-- represents a schedule
type MappedSchedule = Map.Map Timeslot Lesson



time :: Lesson -> Timeslot
time (Lesson {day=day, timeslot=timeslot}) = (day, timeslot)


{-
  Transform a MappedSchedule into a printable,
  and more importantly, readable String
-}
formatSchedule :: MappedSchedule -> String
formatSchedule hours =
  intercalate "\n" $ [header] ++ (map formatDay allHours)

  where
    allHours = [(i, [1..slotsPerDay]) | i <- [1..daysPerWeek]]

    formatLesson :: Timeslot -> String
    formatLesson i =
      printf "%10v" $ maybe [] subject (Map.lookup i hours)

    formatDay :: (Int, [Int]) -> String
    formatDay (i, l) = intercalate " | " [formatLesson (i, j) | j <- l]

    header = printf "Total Weight: %10v" (totalWeight hours)


totalWeight :: MappedSchedule -> Int
totalWeight m = Map.foldl (+) 0 $ Map.map weight m


{-
  Main evaluation function
  Transforms a list of weighted 'Lesson's into a list of lightest schedules
  by branching the evaluation at avery point where there is a timeslot collision
-}
calc :: [Lesson] -> [MappedSchedule]
calc lessons =
  calcStep x lists (Map.empty) minList
  where
    mappedLessons                 = Map.fromListWith (++) $ map (\x -> (subject x, [x])) lessons
    sortedLessons                 = Map.map (List.sortBy (Ord.comparing weight)) mappedLessons
    (minListPrimer, listsValues)  = unzip $ map (\(t, (x : xs)) -> (x, (t, xs))) $ Map.toList sortedLessons
    lists                         = Map.fromList listsValues
    (x : minList)                 = List.sortBy (Ord.comparing weight) minListPrimer


{-
  Helper function for calc
  represents a recusively called and forking calculation step
-}
calcStep :: Lesson -> MappedLessons -> MappedSchedule -> [Lesson] -> [MappedSchedule]
calcStep x lists hourMap minList =
  case (Map.lookup (time x) hourMap) of

    Nothing   ->
      if (null minList)
        then
          [newMap]
        else
          let (c : cs) = minList in
            calcStep c lists newMap cs

    Just old  ->
      let
        r1 = (reduceLists (subject x) lists) hourMap minList
        r2 = (reduceLists (subject old) lists) newMap minList
      in
      r1 ++ r2

  where
    newMap = Map.insert (time x) x hourMap

    reduceLists :: String -> MappedLessons -> (MappedSchedule -> [Lesson] -> [MappedSchedule])
    reduceLists s lists =
      case (Map.lookup s lists) of
        Nothing       -> noResult
        Just []       -> noResult
        Just (c : cs) -> calcStep c (Map.insert s cs lists)
        where
          noResult = \y x -> []
